### PR TITLE
Initial dnatracing refactor

### PR DESCRIPTION
**Pull request by ns-rse**
Work to date on refactoring DNAtracing. This introduces `topostats.tracing.trace_dna` sub-module with a class `traceDNA`.

This is far from complete! None of the code is currently used _anywhere_ in processing, but that is ok, the code can sit there and wait until it is complete with comprehensive tests and then we can switch out the `topostats.tracing.dnatracing` sub-module and its classes with these new ones.

I want to avoid asking people to review lots of code (and there is already a lot in here!) and its important that I get feedback regularly.

* Currently in the classes within `topostats.tracing.dnatracing` the _whole_ image is processed. `traceDNA` class takes a different approach and the unit of processing is now an individual grain. This avoids having to repeatedly loop over all found grains and in turn avoids the issues that @Jean-Du encountered with statistics not aligning. When it comes time to process a whole image each grain is passed in and the results returned (or `None` if it fails to meet thresholds/elsewhere).
* The `dnaTrace` class takes the approach of finding the co-ordinates of each skeletonised molecule and working out if it has an adjacent point to count how many ends there are. I've used a method which leverages NumPy's vectorisation and have tried to explain this in the docstring.
* I'm trying to take a Test Driven Development approach and write tests before the code (or if failing in conjunction with). These are heavily parameterised because there are lots of possibilities for molecules but they use small dummy examples rather than large observed skeletons. Some of these may not (yet) have been observed in real data, but that doesn't preclude them from cropping up in the future. The aim is to test ~all~ as many eventualities as I can think of up-front during development.
* Switched the `get_skeleton()` function to a class.
* `topostats.tracing.utils` sub-module is very much a  work in progress, it holds skeleton classes and methods at the moment.

### CODE REVIEW

**Review by SylviaWhittle — COMMENTED** [review]
Can't see any structural issues. Flagged a typo and a test that might have been accidentally left commented out, though as this is a WIP, it's hard to tell. 

I can't really speak on the tracing front, it seems to make sense to me so far, but I think others who are working on / with dnatracing should have a look.

I've made the following chart for my own understanding as I read through the code, posting it here as it might be helpful (or not, if it's wrong) to other reviewers

image

File: topostats/tracing/skeletonize.py [hunk 1]
```diff
@@ -4,76 +4,138 @@
...
+
+        Parameters
+        ----------
+        method: str
+            Method to use for skeletonizing, methods are 'zhang' (default), 'lee', 'medial_axis', and 'thin'.
+
+        Returns
+        -------
+        Callable
+            Returns the function appropriate for the required skeletonizing method.
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/tracing/skeletonize.py:67]: Good spot, unlike with [thresholding](https://github.com/AFM-SPM/TopoStats/blob/main/topostats/thresholds.py) I've not returned the method so I don't need to return a callable object, I just wanted the result of whichever method.

I've changed this to align with the pattern used in `thresholds.py`.

File: tests/tracing/test_tracedna.py [hunk 2]
```diff
@@ -0,0 +1,1208 @@
...
+        #             [0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
+        #             [0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
+        #             [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
+        #             [0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0],
+        #             [0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0],
+        #             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+        #             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+        #         ]
+        #     ),
+        # ),
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/tracing/test_tracedna.py:526]: I found I got weird, inconsistent results from the medial method which baffled me and so I commented some of the tests out.

I'll look again, but generally I don't think `medial` is a method that should be used for this work anyway and may remove it as an option completely.

**Comment by ns-rse** [comment]
I'm going to close this for now.  There is, I think, some useful refactoring in here (e.g. vectorisation of counting how many adjacent non-zero cells any given point has) but the more I dig into the `topostats/tracing/tracingfuncs.py` the more overwhelmed I'm becoming. There is reimplementation of existing methods in scikit-image with some additional functionality that I need to understand, for example `getDNAmolHeightStats.finalSkeltonisationIteration()` chops off the ends of traces for some reason, not clear why.

Many many functions are done via loops over co-ordinates and I suspect these can be vectorised to improve performance. Its all going to considerably more time than I first anticipated.


I'll be starting a new strategy to...

1. Get unit tests in place for things using a single molecule/grain.
2. Remove the logic that loops over the whole image from `dnatracing.py` and have it handle a single grain. The logic that deals with all grains should be in `topostats.processing.process_scan()`.
3. Understand why certain things are done as they are in `dnatracing`/`tracingfuncs` and then move functionality to this branch.

**Pull request closed without merging**